libcdb: improve the search speed of `search_by_symbol_offsets`

the-soloist opened the pull request:
While using `search_by_symbol_offsets`, I found that the search speed for `build_id` was significantly slower compared to other hash types.

```py
# https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/libcdb.py#L26-L30
HASHES = {
    'build_id': lambda path: enhex(ELF(path, checksec=False).buildid or b''),
    'sha1': sha1filehex,
    'sha256': sha256filehex,
    'md5': md5filehex,
}
```

The reason for this is that ELF loads too many things. I attempted to replace it with `ELFFile`, which noticeably improved the speed, but it introduced redundant functionality. I couldn't think of a simple way to implement it, so I added a `hash_type` parameter to `search_by_symbol_offsets`, with a default setting of `md5` to speed up `search_by_symbol_offsets`, and provide users with a controllable option.

I'm testing on the following code:
```py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
from elftools.elf.elffile import ELFFile
from pwn import *


context.log_level = "info"
context.local_libcdb = "/path/to/libc-database"


def _buildid(path):
    elf = ELFFile(open(path, "rb"))
    section = elf.get_section_by_name('.note.gnu.build-id')
    if section:
        return enhex(section.data()[16:])
    return b""


log.waitfor("searching build_id")
os.system("rm -rf ~/.cache/.pwntools-cache-*")
time_start = time.time()
path = libcdb.search_by_build_id("70a4c953a01ddc232969c27031e7f948338ca137", offline_only=True, unstrip=False)
libc = ELF(path, checksec=False)
print(f"cost {time.time() - time_start}s", libc)


log.waitfor("searching symbol offsets (build_id)")
os.system("rm -rf ~/.cache/.pwntools-cache-*")
time_start = time.time()
path = libcdb.search_by_symbol_offsets({'puts': 0xa30, 'printf': 0x8f0}, offline_only=True, unstrip=False, hash_type="build_id")
libc = ELF(path, checksec=False)
print(f"cost {time.time() - time_start}s", libc)


log.waitfor("searching symbol offsets (md5)")
os.system("rm -rf ~/.cache/.pwntools-cache-*")
time_start = time.time()
path = libcdb.search_by_symbol_offsets({'puts': 0xa30, 'printf': 0x8f0}, offline_only=True, unstrip=False, hash_type="md5")
libc = ELF(path, checksec=False)
print(f"cost {time.time() - time_start}s", libc)


log.success("patch libcdb.HASHES")
libcdb.HASHES["build_id"] = _buildid


log.waitfor("searching build_id (with ELFFile)")
os.system("rm -rf ~/.cache/.pwntools-cache-*")
time_start = time.time()
path = libcdb.search_by_build_id("70a4c953a01ddc232969c27031e7f948338ca137", offline_only=True, unstrip=False)
libc = ELF(path, checksec=False)
print(f"cost {time.time() - time_start}s", libc)


log.waitfor("searching symbol offsets (build_id with ELFFile)")
os.system("rm -rf ~/.cache/.pwntools-cache-*")
time_start = time.time()
path = libcdb.search_by_symbol_offsets({'puts': 0xa30, 'printf': 0x8f0}, offline_only=True, unstrip=False, hash_type="build_id")
libc = ELF(path, checksec=False)
print(f"cost {time.time() - time_start}s", libc)
```

![image](https://github.com/Gallopsled/pwntools/assets/46804083/d169b932-ded8-48a7-9abf-fae13d216ce5)

and found another question https://github.com/Gallopsled/pwntools/issues/2414

the-soloist committed:
Add `hash_type` for `search_by_symbol_offsets`

the-soloist committed:
Add docs

the-soloist committed:
Update CHANGELOG

peace-maker commented:
I think we can avoid walking the local database directory again here in the first place instead. When finding a match in the local libc-database, we know the id and thus the filename of the libc we want to return. Maybe allow the `id` to be searched in `search_by_hash` and special case it in the local_database provider.

the-soloist committed:
Allow search `id` in search_by_hash

the-soloist committed:
Fix py2.7 test

the-soloist commented:
I agree that handling `id` separately within the providers is a good approach, it allows the use of libcdb's caching feature. However, this will cause some variable name to lose its original meaning (it's not hash type). I've tried writing some code, could you give me some suggestions?

Arusekk commented:
I'm not sure I like `hash_type="id"` (maybe `hash_type="filename"` would be better?). I think the build ID should be the default, it should just be parsed quicker, maybe we can have a separate function for extracting build id (at C speed ideally), but come on, reading only the first page of a file *should* be quicker than reading all of it, especially on HDDs; also, build-id does not change if you strip/unstrip or move the file around. If our ELF implementation is a bottleneck, we can resort to implementing separate functionality just for turbofast build-id extraction.

the-soloist committed:
Rename `hash_type` to `search_type`

the-soloist committed:
Rename `TYPES['id']` to `TYPES['libs_id']`

the-soloist committed:
Rename part `hex_encoded_id` to `search_target`

the-soloist committed:
Turbofast extract build id

the-soloist commented:
Sorry too late. I added a `_turbofast_extract_build_id ` function and made some adjustments to the variable names.


```py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
from pwn import *

context.log_level = "info"
context.local_libcdb = "/root/S3cur1ty/libc-database"


log.waitfor("searching build_id")
os.system("rm -rf ~/.cache/.pwntools-cache-*")
time_start = time.time()
path = libcdb.search_by_build_id("6ee9454b96efa9e343f9e8105f2fa4529265ea05", offline_only=True, unstrip=False)
libc = ELF(path, checksec=False)
print(f"cost {time.time() - time_start}s", libc)
```

![image](https://github.com/user-attachments/assets/a705ddc0-1cdd-494b-b759-8b322581217a)

the-soloist committed:
Fix docs

peace-maker commented on pwnlib/libcdb.py:
```diff
@@ -86,16 +117,16 @@ def query_libc_rip(params):
-    params = {hash_type: hex_encoded_id}
+    if search_type == 'build_id':
+        search_type = 'buildid'
+    params = {search_type: search_target}
```
This actually uses `id` as identifier instead of `libs_id`. That's where the name came from. If we invent `libs_id` now, this has to be mapped here.

peace-maker commented on pwnlib/libcdb.py:
```diff
@@ -23,8 +24,35 @@
+    # search NT_GNU_BUILD_ID and b"GNU\x00" (type+name)
+    idx = data.find(bytes.fromhex("03000000 474e5500"))
+    if idx == -1:
+        return enhex(b'')
```
Maybe fallback to proper file parsing here instead?

peace-maker commented on CHANGELOG.md:
```diff
@@ -83,6 +83,7 @@ The table below shows which release corresponds to each branch, and what date th
 - [#2376][2376] Return buffered data on first EOF in tube.readline()
 - [#2387][2387] Convert apport_corefile() output from bytes-like object to string
 - [#2388][2388] libcdb: add `offline_only` to `search_by_symbol_offsets`
+- [#2413][2413] libcdb: improve the search speed of `search_by_symbol_offsets`
```
rebase on latest dev please and move this to the 4.15.0 changelog

the-soloist replied:
I don't know how to rebase just the `CHANGELOG.md`. Do I need open a new PR?

the-soloist committed:
Add a map for types key

the-soloist committed:
Extract proper buildid

the-soloist committed:
Fix docs

peace-maker committed:
Merge branch 'dev' into dev

the-soloist committed:
Fix E0606

peace-maker commented on pwnlib/libcdb.py:
```diff
@@ -661,21 +728,52 @@ def search_by_symbol_offsets(symbols, select_index=None, unstrip=True, return_as
     if return_as_list:
         return [libc['buildid'] for libc in matching_list]
 
+    if search_type in MAP_TYPES.keys():
```
This can be simplified  to
```python
match_type = MAP_TYPES.get(search_type, search_type)
```

peace-maker commented on pwnlib/libcdb.py:
```diff
@@ -661,21 +728,52 @@ def search_by_symbol_offsets(symbols, select_index=None, unstrip=True, return_as
+
+def search_by_libs_id(libs_id, unstrip=True, offline_only=False):
+    """
+    Given a hex-encoded Build ID, attempt to download a matching libc from libcdb.
```
The docs mention a build id instead of the library filename.

peace-maker commented on pwnlib/libcdb.py:
```diff
@@ -661,21 +728,52 @@ def search_by_symbol_offsets(symbols, select_index=None, unstrip=True, return_as
-    return search_by_build_id(selected_libc['buildid'], unstrip=unstrip, offline_only=offline_only)
+    return search_by_hash(selected_libc[match_type], search_type=search_type, unstrip=unstrip, offline_only=offline_only)
+
+def search_by_libs_id(libs_id, unstrip=True, offline_only=False):
```
`search_by_libs_id` needs to be added to the `__all__` array at the bottom of the file.

peace-maker reviewed (changes requested):
Thank you for your persistence while implementing this! I've finally had a minute to look at the current state again and I think it's good minus a few nits. I can fix those myself too if you're busy.

the-soloist committed:
Simplify get `mapped_type`

the-soloist committed:
Add `search_by_libs_id` to `__all__`

the-soloist committed:
Modify docs

the-soloist committed:
Fix py2.7

the-soloist committed:
Fix reference

the-soloist commented:
I’ve addressed the nits you mentioned and have completed the modifications. I hope we can merge this soon so I can move forward with the libcdb-cli changes.

peace-maker committed:
Merge branch 'dev' into dev

peace-maker reviewed (approved):
Cool, thank you!
